lib/deimos/active_record_consume/batch_consumption.rb

```diff
@@ -88,36 +88,73 @@ def update_database(messages)
       def upsert_records(messages)
         key_cols = key_columns(messages)
 
-        # Create payloads with payload + key attributes
-        upserts = messages.map do |m|
-          attrs = if self.method(:record_attributes).parameters.size == 2
-                    record_attributes(m.payload, m.key)
-                  else
-                    record_attributes(m.payload)
-                  end
-
-          attrs&.merge(record_key(m.key))
-        end
-
+        # Create ActiveRecord Models with payload + key attributes
+        upserts = build_models(messages)
         # If overridden record_attributes indicated no record, skip
         upserts.compact!
+        # apply ActiveRecord validations and fetch valid Records
+        valid_upserts = filter_models(upserts)
+
+        save_records_to_database(@klass, key_cols, valid_upserts)
+
+        import_associations(valid_upserts, key_cols) unless @association_list.empty?
+      end
+
+      def save_records_to_database(record_class, key_cols, records)
+        columns = :all
+        if record_class.respond_to?(:bulk_import_columns)
+          columns = record_class.bulk_import_columns
```

Comment:
I'd want this to be a configuration for the consumer, not the record class. Let's try to keep the model class itself clean. Probably the easiest way is just to have a method on the consumer that takes a class and returns the columns.

Reply:
the nice thing about the method is that it can be whatever you want it to be as long as it returns an array of column names :)

Reply:
Done. I tested this in RSpecs as well